replace O(n^2) LOC with constant-prop in Dates

LilithHafner opened the pull request:
```julia
julia> using Dates, Base.Order

julia> @code_llvm lt(By(Dates.days ∘ oneunit, Reverse), Week(3), Quarter(2))
;  @ ordering.jl:119 within `lt`
define i8 @julia_lt_9867([1 x i64]* nocapture noundef nonnull readonly align 8 dereferenceable(8) %0, [1 x i64]* nocapture noundef nonnull readonly align 8 dereferenceable(8) %1) #0 {
top:
  ret i8 0
}
```

Commit:
replace O(n^2) LOC with constant-prop in Dates

quinnj commented on stdlib/Dates/src/periods.jl:
```diff
@@ -170,7 +131,7 @@ struct CompoundPeriod <: AbstractTime
         n = length(p)
         if n > 1
-            sort!(p, rev=true, lt=periodisless)
+            sort!(p, rev = true, by = days ∘ oneunit)
```
Hmmm, what's going on here? Can you explain the magic?

LilithHafner replied:
Sure!

We sort periods in decreasing order (`rev = true`) according to the length of the _period_ (e.g `Week(1)` before `Hour(240)`). There are many ways to get the length of a period `p` and `days(oneunit(p))` seems like a reasonable choice. How many days is one `Second`/`Week`/`Year`? This is sometimes a `Flaot64` and sometimes an `Int64` but we already don't have type stability because we are working with `Vector{Period}` and this constant folds once the type is known.

LilithHafner replied:
That explanation feels a bit long to include as a comment, but we could if you think it is appropriate.

gbaraldi replied:
Yeah adding a comment on how this works is probably a good idea, specially given that the code it replaces is pretty obvious in how it works

LilithHafner replied:
Does this explain it adequitely?
```
# We sort periods in decreasing order (rev = true) according to the length of
# the period's type (by = days ∘ oneunit). We sort by type, not value, so that
# we can merge equal types.
#
# This works by computing how many days are in a single period, and then sorting
# by that. For example, (days ∘ oneunit)(Week(10)) = days(oneunit(Week(10))) =
# days(Week(1)) = 7, which is less than (days ∘ oneunit)(Month(-2)) = 30.436875
```

LilithHafner replied:
Even if this is less obvious, I think its a worthwhile change because it takes the number of new methods to be implemented for a new `Period` type to support `CompundPeriod`s down from O(n) to 0.

gbaraldi replied:
That sounds quite clear 👍

LilithHafner replied:
Thanks!

Commit:
add brief explanation

Commit:
expand explanation

LilithHafner committed:
Merge branch 'JuliaLang:master' into cleanup-dates-1

LilithHafner committed:
Switch from days to nanoseconds

To distinguish small period types from each other. 
This avoids #46150 and, more importantly, fixes:

```julia
julia> Dates.days(Millisecond(1))
0

julia> Dates.days(Second(1))
0
```

KristofferC committed:
Merge branch 'master' into cleanup-dates-1
